File: node/13_use_built-in_roles.js
const { MilvusClient } = require("@zilliz/milvus2-sdk-node")

const address = "YOUR_CLUSTER_ENDPOINT"
const token = "YOUR_CLUSTER_TOKEN"

// Include the following in an async function declaration
async function main () {

    // 1. Connect to the cluster
    const client = new MilvusClient({address, token})

    // 2. Create user
    const users1 = await client.listUsers()

    if (!users1.usernames.includes("user1")) {
        await client.createUser({
            username: "user1",
            password: "P@ssw0rd!",
        })
    }

    // 3. Update user credentials

    await client.updateUser({
        username: "user1",
        oldPassword: "P@ssw0rd!",
        newPassword: "P@ssw0rd!!",
    })

    // 4. List users

    const users = await client.listUsers()

    console.log(users.usernames)

    // Output
    // 
    // [ 'db_admin', 'user1' ]
    // 

    // 5. Assigne role to user

    await client.addUserToRole({
        username: "user1",
        rolename: "db_ro",
    })

    // 6. List roles

    const roles = await client.listRoles()

    console.log(roles.results)

    // Output
    // 
    // [
    //   { users: [ [Object] ], role: { name: 'admin' } },
    //   { users: [ [Object] ], role: { name: 'db_admin' } },
    //   { users: [], role: { name: 'db_ro' } },
    //   { users: [], role: { name: 'db_rw' } },
    //   { users: [], role: { name: 'public' } }
    // ]
    // 

    // 7. Remove user from role

    await client.removeUserFromRole({
        username: "user1",
        rolename: "db_ro",
    })

    // 7. Delete user

    await client.deleteUser({
        username: "user1",
    })

}

main()

File: python/13_use_built-in_roles.py
import json, os, time
from pymilvus import connections, Role, utility

CLUSTER_ENDPOINT="YOUR_CLUSTER_ENDPOINT" # Set your cluster endpoint
TOKEN="YOUR_CLUSTER_TOKEN" # Set your token

connections.connect(
    alias='default', 
    #  Public endpoint obtained from Zilliz Cloud
    uri=CLUSTER_ENDPOINT,
    # API key or a colon-separated cluster username and password
    token=TOKEN, 
)

# 1. Create user

if not 'user1' in utility.list_usernames():
    utility.create_user(user='user1', password='P@ssw0rd!')

# 2. Update user credentials

utility.update_password(
    user='user1',
    old_password='P@ssw0rd!',
    new_password='P@ssw0rd!!'
)

# 3. List users

print(utility.list_usernames())

# Output
#
# ["db_admin", "user1"]



userInfo = utility.list_users(include_role_info=True)

users = [ { "user": u.username, "roles": list(u.roles) } for u in userInfo.groups ]

print(users)

# Output
#
# [
#     {
#         "user": "db_admin",
#         "roles": [
#             "db_admin"
#         ]
#     },
#     {
#         "user": "user1",
#         "roles": []
#     }
# ]



# 4. Assign role

role = Role("db_ro")

role.add_user("user1")

# 5. Get users of a specific role

users = list(role.get_users())

print(users)

# Output
#
# ["user1"]



# 6. List roles

roleInfo = utility.list_roles(include_user_info=True)

roles = [ { "role": g.role_name, "users": list(g.users) } for g in roleInfo.groups ]

print(roles)

# Output
#
# [
#     {
#         "role": "db_admin",
#         "users": [
#             "db_admin"
#         ]
#     },
#     {
#         "role": "db_ro",
#         "users": [
#             "user1"
#         ]
#     },
#     {
#         "role": "db_rw",
#         "users": []
#     }
# ]



# 7. Remove user from role

role.remove_user("user1")

# 8. Delete user
utility.delete_user('user1')